core/src/main/java/io/github/thibaultbee/streampack/internal/encoders/mediacodec/MediaCodecEncoder.kt

```diff
@@ -0,0 +1,493 @@
+                mediaCodec.flush()
+            } else {
+                mediaCodec.stop()
+            }
+
+        } catch (e: IllegalStateException) {
+            Logger.d(tag, "Not running")
+        } finally {
+            isStopped = true
+        }
+    }
+
+    override suspend fun stopStream() {
+        withContext(dispatcher) {
+            stopStreamSync()
+        }
+    }
+
+    private fun releaseSync() {
+        if (isReleased) {
+            return
+        }
+        try {
+            mediaCodec.release()
+
+            if (input is SurfaceInput) {
+                input.release()
+            }
+        } catch (_: Exception) {
+        } finally {
+            isReleased = true
+        }
+    }
+
+    override fun release() {
+        runBlocking {
+            if (!isStopped) {
+                stopStream()
+            }
+            withContext(dispatcher) {
+                releaseSync()
+            }
+        }
+    }
+
+    private fun notifyError(e: Exception) {
+        var listener: IEncoder.IEncoderListener
+        var listenerExecutor: Executor
+        synchronized(listenerLock) {
+            listener = this.listener
+            listenerExecutor = this.listenerExecutor
+        }
+        listenerExecutor.execute {
+            listener.onError(e)
+        }
+    }
+
+    private fun handleError(e: Exception) {
+        isOnError = true
+        if (!isStopped) {
+            encoderExecutor.execute {
+                stopStreamSync()
+                notifyError(e)
+            }
+        } else {
+            notifyError(e)
+        }
+    }
+
+    private fun reachEndOfStream() {
+        runBlocking {
+            stopStream()
+        }
+    }
+
+    private inner class EncoderCallback : MediaCodec.Callback() {
+        override fun onOutputBufferAvailable(
+            codec: MediaCodec,
+            index: Int,
+            info: BufferInfo
+        ) {
+            encoderExecutor.execute {
+                if (isStopped) {
+                    Logger.w(tag, "Receives frame after codec is reset.")
+                    return@execute
+                }
+                if (isOnError) {
+                    return@execute
+                }
+
+                if (isBufferInfoValid(info, tag)) {
+                    try {
+                        var listener: IEncoder.IEncoderListener
+                        var listenerExecutor: Executor
+                        synchronized(listenerLock) {
+                            listener = this@MediaCodecEncoder.listener
+                            listenerExecutor = this@MediaCodecEncoder.listenerExecutor
+                        }
+
+                        try {
+                            Frame(
+                                codec,
+                                index,
+                                info
+                            ).let { frame ->
+                                listenerExecutor.execute {
```

Comment:
Its hard to imagine that executing through two executors isn't going to introduce latency

Reply:
Just a reminder that before, these callbacks were executed on a Handler (same as `encoderExecutor`). That leaves just one executor.
If `inOutputFrame` takes a long time, I don't want to block encoder thread.
I remember that you introduced a thread in the RTMP publisher, you won't have to do it.